intset_test.go

```diff
@@ -0,0 +1,212 @@
+
+	intersect := fullSet.Intersection(firstQuartile)
+	assert.True(t, intersect.EqualSet(firstQuartile))
+	assert.Equal(t, firstQuartile.count, intersect.count)
+
+	intersect = firstQuartile.Intersection(thirdQuartile)
+	assert.True(t, intersect.IsEmpty())
+	assert.Equal(t, 0, intersect.count)
+
+	intersect = secondToFifthDecile.Intersection(firstQuartile)
+	distinctNums := len(getDistinctInts(arr[int(len(arr)/5):int(len(arr)/4)]))
+	for i := intersect.Range(); i.Next(); {
+		assert.True(t, secondToFifthDecile.Has(i.Value()) && firstQuartile.Has(i.Value()))
+	}
+	assert.Equal(t, distinctNums, intersect.count)
+}
+
+func TestIntSetUnion(t *testing.T) {
+	t.Parallel()
+
+	arr, fullSet := generateIntArrayAndSet()
+	set := NewIntSet(arr[:len(arr)/2]...)
+	distinct := getDistinctInts(arr)
+	for _, i := range arr[len(arr)/2:] {
+		assert.False(t, set.Has(i))
+	}
+
+	union := set.Union(fullSet)
+	for _, i := range distinct {
+		assert.True(t, union.Has(i))
+	}
+	assert.Equal(t, len(distinct), union.count)
+
+	union = fullSet.Union(set)
+	for _, i := range distinct {
+		assert.True(t, union.Has(i))
+	}
+	assert.Equal(t, len(distinct), union.count)
+}
+
+func TestIntSetIsSubsetOf(t *testing.T) {
+	t.Parallel()
+
+	arr, fullSet := generateIntArrayAndSet()
+	assert.True(t, NewIntSet().IsSubsetOf(fullSet))
+	assert.True(t, fullSet.IsSubsetOf(fullSet))
+	assert.True(t, NewIntSet(arr[:len(arr)/2]...).IsSubsetOf(fullSet))
+	assert.False(t, NewIntSet(arr[:len(arr)/2]...).IsSubsetOf(NewIntSet(arr[int(len(arr)/3):]...)))
+}
+
+func TestIntSetWhere(t *testing.T) {
+	t.Parallel()
+
+	arr, fullSet := generateIntArrayAndSet()
+	var evens, odds []int
+	evenPred := func(e int) bool { return e%2 == 0 }
+	oddPred := func(e int) bool { return e%2 != 0 }
+
+	for _, i := range arr {
+		if evenPred(i) {
+			evens = append(evens, i)
+		} else {
+			odds = append(odds, i)
+		}
+	}
+
+	evenSet := NewIntSet(evens...)
+	oddSet := NewIntSet(odds...)
+
+	evenPredWhere := fullSet.Where(evenPred)
+	oddPredWhere := fullSet.Where(oddPred)
+	assert.True(t, evenPredWhere.EqualSet(evenSet))
+	assert.Equal(t, evenSet.count, evenPredWhere.count)
+	assert.True(t, oddPredWhere.EqualSet(oddSet))
+	assert.Equal(t, oddSet.count, oddPredWhere.count)
+	assert.True(t, NewIntSet().Where(evenPred).EqualSet(NewIntSet()))
+}
+
+func TestIntSetMap(t *testing.T) {
+	t.Parallel()
+
+	subtract := func(e int) int { return e - 1 }
+	arr, fullSet := generateIntArrayAndSet()
+	mappedArr := make([]int, 0, len(arr))
+
+	for _, i := range arr {
+		mappedArr = append(mappedArr, subtract(i))
+	}
+
+	mappedSet := NewIntSet(mappedArr...)
+
+	assert.True(t, mappedSet.EqualSet(fullSet.Map(subtract)))
+	assert.True(t, NewIntSet().EqualSet(NewIntSet().Map(subtract)))
+}
+
+func generateIntArrayAndSet() ([]int, IntSet) {
+	arr := make([]int, 0, maxIntArrLen)
+	curr := float64(1.0)
+	for i := 1; i < maxIntArrLen; i++ {
+		arr = append(arr, int(curr))
+		curr *= math.Pow(2, 64/math.Pow(10, 6))
```

Comment:
The multiplier is a constant. Calculate it once outside the function.

Reply:
added